packages/yew/src/functional/hooks/mod.rs

```diff
@@ -23,53 +56,49 @@ use std::rc::Rc;
-///
-/// See the pre-defined hooks for examples of how to use this function.
-///
-/// [Yew Docs]: https://yew.rs/next/concepts/function-components/custom-hooks
-pub fn use_hook<InternalHook: 'static, Output, Tear: FnOnce(&mut InternalHook) + 'static>(
-    initializer: impl FnOnce() -> InternalHook,
-    runner: impl FnOnce(&mut InternalHook, HookUpdater) -> Output,
-    destructor: Tear,
-) -> Output {
-    if !CURRENT_HOOK.is_set() {
-        panic!("Hooks can only be used in the scope of a function component");
+pub(crate) fn use_hook<'hook, T, O>(
+    initializer: impl 'hook + FnOnce() -> T,
+    runner: impl 'hook + FnOnce(&mut T, HookUpdater) -> O,
+) -> impl 'hook + Hook<Output = O>
+where
+    T: 'static,
+    O: 'hook,
+{
+    struct HookProvider<'a, T, O> {
+        initializer: Box<dyn FnOnce() -> T + 'a>,
+        runner: Box<dyn FnOnce(&mut T, HookUpdater) -> O + 'a>,
     }
 
-    // Extract current hook
-    let updater = CURRENT_HOOK.with(|hook_state| {
-        // Determine which hook position we're at and increment for the next hook
-        let hook_pos = hook_state.counter;
-        hook_state.counter += 1;
-
-        // Initialize hook if this is the first call
-        if hook_pos >= hook_state.hooks.len() {
-            let initial_state = Rc::new(RefCell::new(initializer()));
-            hook_state.hooks.push(initial_state.clone());
-            hook_state.destroy_listeners.push(Box::new(move || {
-                destructor(initial_state.borrow_mut().deref_mut());
-            }));
-        }
+    impl<T, O> Hook for HookProvider<'_, T, O>
+    where
+        T: 'static,
+    {
+        type Output = O;
 
-        let hook = hook_state
-            .hooks
-            .get(hook_pos)
-            .expect("Not the same number of hooks. Hooks must not be called conditionally")
-            .clone();
+        fn run(self, ctx: &mut HookContext) -> Self::Output {
+            let Self {
+                initializer,
+                runner,
+            } = self;
 
-        HookUpdater {
-            hook,
-            process_message: hook_state.process_message.clone(),
+            // Extract current hook
+            let updater = ctx.next_state(initializer);
+
+            // Execute the actual hook closure we were given. Let it mutate the hook state and let
+            // it create a callback that takes the mutable hook state.
+            let mut hook = updater.borrow_mut::<T>();
+
+            runner(&mut *hook, updater.clone())
         }
-    });
+    }
 
-    // Execute the actual hook closure we were given. Let it mutate the hook state and let
-    // it create a callback that takes the mutable hook state.
-    let mut hook = updater.hook.borrow_mut();
-    let hook: &mut InternalHook = hook
-        .downcast_mut()
-        .expect("Incompatible hook type. Hooks must always be called in the same order");
+    HookProvider {
+        initializer: Box::new(initializer),
+        runner: Box::new(runner),
```

Comment:
Why the boxes here?

Reply:
`use_hook` is actually subject to removal in a upcoming pull request so it does not really matter what kind of implementation is used here.